File: src/main/java/backend/service/ClassAnnouncementService.java
package backend.service;

import backend.domain.Admin;
import backend.domain.ClassAnnouncement;
import backend.domain.ClassEntity;
import backend.repository.ClassAnnouncementDBRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Service
public class ClassAnnouncementService {

    private final ClassAnnouncementDBRepository classAnnouncementRepository;

    @Autowired
    public ClassAnnouncementService(ClassAnnouncementDBRepository classAnnouncementRepository) {
        this.classAnnouncementRepository = classAnnouncementRepository;
    }

    // Create a new announcement
    @Transactional
    public ClassAnnouncement createAnnouncement(Admin admin, ClassEntity classEntity, String message, LocalDate date) {
        // Validare date de intrare
        if (admin == null) {
            throw new IllegalArgumentException("Admin cannot be null");
        }
        if (classEntity == null) {
            throw new IllegalArgumentException("Class cannot be null");
        }
        if (message == null || message.trim().isEmpty()) {
            throw new IllegalArgumentException("Message cannot be empty");
        }
        if (date == null) {
            throw new IllegalArgumentException("Date cannot be null");
        }

        ClassAnnouncement announcement = new ClassAnnouncement(admin, classEntity, message, date);
        return classAnnouncementRepository.save(announcement);
    }

    // Find announcement by ID
    public Optional<ClassAnnouncement> getAnnouncementById(Long id) {
        if (id == null) {
            throw new IllegalArgumentException("ID cannot be null");
        }
        return classAnnouncementRepository.findById(id);
    }

    // Find all announcements
    public List<ClassAnnouncement> getAllAnnouncements() {
        return classAnnouncementRepository.findAll();
    }

    // Find announcements for a specific class
    public List<ClassAnnouncement> getAnnouncementsByClassId(Long classId) {
        if (classId == null) {
            throw new IllegalArgumentException("Class ID cannot be null");
        }
        return classAnnouncementRepository.findByClassEntityId(classId);
    }

    // Find announcements by admin ID
    public List<ClassAnnouncement> getAnnouncementsByAdminId(Long adminId) {
        if (adminId == null) {
            throw new IllegalArgumentException("Admin ID cannot be null");
        }
        return classAnnouncementRepository.findByAdminId(adminId);
    }

    // Find announcements for a class after a specific date
    public List<ClassAnnouncement> getAnnouncementsForClassAfterDate(Long classId, LocalDate date) {
        if (classId == null || date == null) {
            throw new IllegalArgumentException("Class ID and date cannot be null");
        }
        return classAnnouncementRepository.findByClassEntityIdAndDateAfter(classId, date);
    }

    // Get announcements for a class sorted by date descending
    public List<ClassAnnouncement> getAnnouncementsByClassIdSortedByDate(Long classId) {
        if (classId == null) {
            throw new IllegalArgumentException("Class ID cannot be null");
        }
        return classAnnouncementRepository.findByClassEntityIdOrderByDateDesc(classId);
    }

    // Update an announcement
    @Transactional
    public ClassAnnouncement updateAnnouncement(Long id, String message, LocalDate date) {
        if (id == null) {
            throw new IllegalArgumentException("ID cannot be null");
        }

        Optional<ClassAnnouncement> existingAnnouncement = classAnnouncementRepository.findById(id);
        if (existingAnnouncement.isEmpty()) {
            throw new IllegalArgumentException("Announcement with ID " + id + " not found");
        }

        ClassAnnouncement announcement = existingAnnouncement.get();

        if (message != null && !message.trim().isEmpty()) {
            announcement.setMessage(message);
        }

        if (date != null) {
            announcement.setDate(date);
        }

        return classAnnouncementRepository.save(announcement);
    }

    // Delete an announcement
    @Transactional
    public void deleteAnnouncement(Long id) {
        if (id == null) {
            throw new IllegalArgumentException("ID cannot be null");
        }

        if (!classAnnouncementRepository.existsById(id)) {
            throw new IllegalArgumentException("Announcement with ID " + id + " not found");
        }

        classAnnouncementRepository.deleteById(id);
    }
}

File: src/main/frontend/src/services/classAnnouncementService.ts
import api from './api.ts';
import { ClassAnnouncementDTO, CreateClassAnnouncementRequest, UpdateClassAnnouncementRequest } from '../types/classAnnouncement';

export const classAnnouncementService = {
    /*
        POST /api/v1/announcements
        Create new announcement
    */
    create: async (data: CreateClassAnnouncementRequest) => {
        return api.post<ClassAnnouncementDTO>('v1/announcements', data);
    },

    /*
        GET /api/v1/announcements
        Get ALL the annoucements
    */
    getAll: async () => {
        return api.get<ClassAnnouncementDTO[]>('v1/announcements');
    },

    /*
        GET /api/v1/announcements/{id}
        Get an announcement by ID
    */
    getById: async (id: number) => {
        return api.get<ClassAnnouncementDTO>(`v1/announcements/${id}`);
    },

    /*
        GET /api/v1/announcements/class/{classId}
        Get announcements for a specific class
    */
    getByClassId: async (classId: number) => {
        return api.get<ClassAnnouncementDTO[]>(`v1/announcements/class/${classId}`);
    },

    /*
        GET /api/v1/announcements/admin/{adminId}
        Get announcements posted by a specific admin
    */
    getByAdminId: async (adminId: number) => {
        return api.get<ClassAnnouncementDTO[]>(`v1/announcements/admin/${adminId}`);
    },

    /*
        PUT /api/v1/announcements/{id}
        Update announcemend  
    */
    update: async (id: number, data: UpdateClassAnnouncementRequest) => {
        return api.put<ClassAnnouncementDTO>(`v1/announcements/${id}`, data);
    },

    /*
        DELETE /api/v1/announcements/{id}
        Delete an announcement (by ID)
    */
    delete: async (id: number) => {
        return api.delete<void>(`v1/announcements/${id}`);
    }
};